airbyte-integrations/connectors/source-netsuite-odbc/source_netsuite_odbc/streams.py

```diff
@@ -0,0 +1,424 @@
+      stream_last_values = current_stream_state['last_values_seen']
+      new_values = {}
+      for key in self.primary_key_column:
+        stream_value = stream_last_values[key] if key in stream_last_values else STARTING_PRIMARY_KEY_VALUE
+        non_stream_value = self.primary_key_last_value_seen[key] if key in self.primary_key_last_value_seen else STARTING_PRIMARY_KEY_VALUE
+        new_values[key] = max(stream_value, non_stream_value)
+      return new_values
+
+    # END STATE MANAGEMENT FUNCTIONS
+
+    # QUERY GENERATION FUNCTIONS
+    def has_composite_primary_key(self):
+      return len(self.primary_key_column) > 1
+    
+    def generate_query(self, sync_mode, stream_slice):
+      if sync_mode == SyncMode.full_refresh:
+        return self.generate_full_refresh_query(stream_slice)
+      elif sync_mode == SyncMode.incremental:
+        return self.generate_incremental_query(stream_slice)
+      else:
+        raise Exception(f'Unsupported Sync Mode: {sync_mode}.  Please use either "full_refresh" or "incremental"')
+
+    # How does full refresh work?
+    # We first split things into stream slices.  Then, we start with filtering our SQL query by > the start date of the stream slice
+    # and <= the end date of the stream slice.  We then sort by the primary key column (s).
+    #  As we fetch records, we start filtering by > the most recent primary key value we've seen.  Because primary keys 
+    # can be composite, we need to handle the cases where the first column might have multiple rows for a single value (since the second
+    # or third row might have different values)
+    # Note:  Outside of the stream slice dates, we don't filter by dates in full refresh.  We instead rely on Netsuite's ids being incremental.
+    def generate_full_refresh_query(self, stream_slice):
+      values = ', '.join(self.properties) # we use values instead of '*' because we want to preserve the order of the columns
+      incremental_column_sorter = f", {self.incremental_column} ASC" if self.incremental_column else ""
+      # per https://docs.oracle.com/en/cloud/saas/netsuite/ns-online-help/section_156257805177.html#subsect_156330163819
+      # date literals need to be wrapped in a to_date function
+      incremental_column_filter = self.generate_incremental_filter_for_full_refresh(stream_slice=stream_slice)
+      primary_key_filter = self.generate_primary_key_filter()
+      primary_key_sorter = self.generate_primary_key_sorter()
+      and_connector = " AND " if primary_key_filter != "" and incremental_column_filter != "" else ""
+      where_clause = "WHERE " if primary_key_filter != "" or incremental_column_filter != "" else ""
+      query = f"""
+        SELECT TOP {NETSUITE_PAGINATION_INTERVAL} {values} FROM {self.table_name} 
+        {where_clause}{primary_key_filter}{and_connector}{incremental_column_filter} 
+        ORDER BY {primary_key_sorter}""" + incremental_column_sorter
```

Comment:
is there a reason full refresh is sorted [differently from incremental](https://github.com/airbytehq/airbyte/pull/32968/files#diff-ec237bc9a560edd56705d285e834a64adfe4cc092ff3356a74349d2f2215adc2R310)? I'd expect this to be sorted by `(incremental_column, pk)`

Reply:
Yes.  The full refresh relies on primary key sort to do a full table scan.  This is because many tables don't even have an incremental column, so we can't rely on it.  Happy to sync on this if needed, but this is intentional.